src/backend/distributed/planner/multi_join_order.c

```diff
@@ -1232,13 +1238,26 @@ LocalJoin(JoinOrderNode *currentJoinNode, TableEntry *candidateTable,
 	joinOnPartitionColumns = JoinOnColumns(currentPartitionColumn,
 										   candidatePartitionColumn,
 										   applicableJoinClauses);
-	if (joinOnPartitionColumns)
+	if (!joinOnPartitionColumns)
 	{
-		nextJoinNode = MakeJoinOrderNode(candidateTable, LOCAL_PARTITION_JOIN,
-										 currentPartitionColumn,
-										 currentPartitionMethod);
+		return NULL;
+	}
+
+	/* shard interval lists must have 1-1 matching for local joins */
+	coPartitionedTables = CoPartitionedTables(
```

Comment:
Well, the function comment used to be wrong, but, with this change it seems to fit into the comment 😄